File: golang/architecture-shell-cqrs/behaviors/unitofwork_behavior.go
package behaviors

import (
	"context"
	"fmt"
	"log/slog"
	"time"

	cqrs "github.com/universalddd/architecture-shell-cqrs"
)

// IResult is a minimal interface for detecting business failures vs infrastructure failures
//
// This is compatible with Architecture.Core Result monad.
// Implementations should use the full Result type from Architecture.Core.
type IResult interface {
	IsFailure() bool
}

// UnitOfWorkBehavior is a pipeline behavior that manages transaction lifecycle
//
// Opens transaction for commands, commits on success (including Result.Failure business errors),
// rolls back on exceptions. Skips transaction management for queries.
// Reuses active transaction for nested commands.
//
// Recommended order: 30 (after validation and authorization)
//
// Transaction semantics per BR-006, BR-007, BR-008:
//   - BeginTransaction(): Opens new transaction if none active; panics if transaction provider unavailable (fail fast)
//   - Nested commands reuse active transaction (HasActiveTransaction check)
//   - Result.Failure() commits transaction (business rejection is valid state per BR-008)
//   - Panics rollback transaction (infrastructure errors)
//   - Logs TransactionId before rollback for correlation (per BR-002)
//
// Example:
//
//	behavior := &UnitOfWorkBehavior{UnitOfWork: myUnitOfWork, Logger: myLogger}
//	mediator := NewMediator(handlers, []interface{}{behavior})
type UnitOfWorkBehavior struct {
	UnitOfWork cqrs.UnitOfWork
	Logger     *slog.Logger
}

// Handle executes the unit of work behavior
func (b *UnitOfWorkBehavior) Handle(ctx context.Context, request interface{}, next func() (interface{}, error)) (interface{}, error) {
	// Skip transaction management for queries (BR-003)
	if !b.isCommand(request) {
		return next()
	}

	// Reuse active transaction for nested commands (BR-005)
	if b.UnitOfWork.HasActiveTransaction() {
		return next()
	}

	// Begin transaction for commands
	startTime := time.Now()

	// Defer rollback on panic (infrastructure error)
	defer func() {
		if r := recover(); r != nil {
			// Log error with TransactionId before rollback per BR-002
			// This ensures correlation even if rollback itself fails
			duration := time.Since(startTime)
			if b.Logger != nil {
				b.Logger.Error("Request failed - rolling back transaction",
					"requestType", fmt.Sprintf("%T", request),
					"transactionId", b.UnitOfWork.TransactionID(),
					"duration", duration.Milliseconds(),
					"error", r)
			}

			// Rollback transaction
			_ = b.UnitOfWork.Rollback(ctx)

			// Re-panic to propagate error
			panic(r)
		}
	}()

	// Begin transaction
	if err := b.UnitOfWork.BeginTransaction(ctx); err != nil {
		// Transaction provider failure - fail fast per BR-006
		return nil, err
	}

	// Execute handler
	response, err := next()

	if err != nil {
		// Log error with TransactionId before rollback per BR-002
		duration := time.Since(startTime)
		if b.Logger != nil {
			b.Logger.Error("Request failed - rolling back transaction",
				"requestType", fmt.Sprintf("%T", request),
				"transactionId", b.UnitOfWork.TransactionID(),
				"duration", duration.Milliseconds(),
				"error", err)
		}

		// Rollback on error
		_ = b.UnitOfWork.Rollback(ctx)
		return nil, err
	}

	// Commit transaction on success or business failure (BR-008)
	// Result.Failure() indicates business validation failure, not infrastructure error
	if err := b.UnitOfWork.Commit(ctx); err != nil {
		return nil, err
	}

	return response, nil
}

// Order returns the execution order for this behavior
func (b *UnitOfWorkBehavior) Order() int {
	return 30
}

// isCommand checks if request is a command
func (b *UnitOfWorkBehavior) isCommand(request interface{}) bool {
	_, ok := request.(interface{ IsCommand() })
	return ok
}

File: java-spring/architecture-shell-cqrs/src/main/java/com/architecture/shell/cqrs/behaviors/UnitOfWorkBehavior.java
package com.architecture.shell.cqrs.behaviors;

import com.architecture.core.functional.Result;
import com.architecture.shell.cqrs.PipelineBehavior;
import com.architecture.shell.cqrs.RequestHandlerDelegate;
import com.architecture.shell.cqrs.UnitOfWork;
import com.architecture.shell.cqrs.BaseRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Pipeline behavior that manages transaction boundaries for commands.
 * Opens transactions, commits on success, rolls back on exceptions.
 *
 * Requirements:
 * - BR-001: Commands execute within transactions
 * - BR-003: Queries skip transaction management
 * - BR-005: Nested commands reuse active transaction
 * - BR-006: Fail fast when transaction provider unavailable
 * - BR-007/BR-008: Result.failure() commits, exceptions rollback
 * - data-model.md:L491-L516: Log TransactionId BEFORE rollback
 *
 * Transaction Semantics:
 * - Command + no active tx → begin, execute, commit/rollback
 * - Command + active tx → reuse, execute, outer command commits/rolls back
 * - Query → skip transaction logic entirely
 * - Result.failure() → COMMIT (business rejection is valid state)
 * - Exception thrown → ROLLBACK (infrastructure failure)
 *
 * DDD Layer: Application Layer Infrastructure
 */
public class UnitOfWorkBehavior<TRequest extends BaseRequest, TResponse>
    implements PipelineBehavior<TRequest, TResponse> {

    private static final Logger logger = LoggerFactory.getLogger(UnitOfWorkBehavior.class);

    private final UnitOfWork unitOfWork;

    public UnitOfWorkBehavior(UnitOfWork unitOfWork) {
        this.unitOfWork = unitOfWork;
    }

    @Override
    public TResponse handle(TRequest request, RequestHandlerDelegate<TResponse> next) {
        // Skip transaction for queries (BR-003)
        if (!(request instanceof com.architecture.shell.cqrs.Command)) {
            return next.handle();
        }

        // Check if transaction already active (nested command scenario per BR-005)
        boolean isOuterCommand = !unitOfWork.hasActiveTransaction();

        if (isOuterCommand) {
            // Outer command: open new transaction (may throw per BR-006 fail-fast)
            try {
                unitOfWork.beginTransaction();
                logger.debug("Transaction started: TransactionId={}", unitOfWork.getTransactionId());
            } catch (Exception ex) {
                logger.error("Failed to begin transaction (BR-006 fail-fast): {}", ex.getMessage());
                throw ex; // Propagate to caller
            }
        } else {
            // Nested command: reuse existing transaction
            logger.debug("Reusing active transaction: TransactionId={}", unitOfWork.getTransactionId());
        }

        try {
            // Execute handler
            TResponse response = next.handle();

            // Only commit if this is the outer command
            if (isOuterCommand) {
                // Check if response is Result.failure() - still commits per BR-007/BR-008
                if (response instanceof Result<?> result && result.isFailure()) {
                    logger.debug("Command returned Result.failure() - committing transaction (business rejection): TransactionId={}",
                        unitOfWork.getTransactionId());
                } else {
                    logger.debug("Command succeeded - committing transaction: TransactionId={}",
                        unitOfWork.getTransactionId());
                }
                unitOfWork.commit();
            }

            return response;

        } catch (Exception ex) {
            // Only rollback if this is the outer command
            if (isOuterCommand) {
                // Log BEFORE rollback per data-model.md:L491-L516
                logger.error("Command failed - TransactionId={} RequestType={} Duration={}ms Exception={}",
                    unitOfWork.getTransactionId(),
                    request.getClass().getSimpleName(),
                    "N/A", // Duration tracked by TelemetryBehavior
                    ex.getMessage(),
                    ex);

                try {
                    unitOfWork.rollback();
                    logger.debug("Transaction rolled back: TransactionId={}", unitOfWork.getTransactionId());
                } catch (Exception rollbackEx) {
                    logger.error("Rollback failed for TransactionId={}: {}",
                        unitOfWork.getTransactionId(),
                        rollbackEx.getMessage(),
                        rollbackEx);
                    // Don't throw rollback exception - preserve original exception
                }
            }
            throw ex; // Propagate original exception
        }
    }
}
